Full Site Editing / 2022 Compatibility

eri-trabiccolo opened the pull request:
## Description

Fixes #1824 and adds FSE compatibilty.
run `npm run build:styles` prior to test 2022 compatiblity (for the forms).

Requires: https://github.com/gocodebox/lifterlms-blocks/pull/146

Additional tests need to be added.
There are at least two knowns "issues" at the moment that might be improved:
1) When with a FSE theme, any notice print at `llms_loop` on a non-singular template results in the notice be printed *before* the `html` opening tag. I was thinking about handling the notices printing in js, at least with FSE, trying to guess the first available element to which append the notice.
2) When a plugin block template is modified in the FSE, the "blue dot", which better shows the template is customized doesn't appear over it. This seems to be a WP bug to me, so I haven't investigated further.

## How has this been tested?
manually

## Screenshots <!-- if applicable -->

## Types of changes
New feature (non-breaking change which adds functionality)

## Checklist:
- [x] My code has been tested.
- [x] My code passes all existing automated tests. <!-- Check code: `composer run-script tests-run`, Guidelines: https://github.com/gocodebox/lifterlms/blob/trunk/tests/README.md -->
- [x] My code follows the LifterLMS Coding & Documentation Standards. <!-- Check code: `composer run-script check-cs-errors`, Guidelines: https://github.com/gocodebox/lifterlms/blob/trunk/docs/coding-standards.md and https://github.com/gocodebox/lifterlms/blob/trunk/docs/documentation-standards.md -->

eri-trabiccolo committed:
initial work

eri-trabiccolo committed:
better templates handling

eri-trabiccolo committed:
admin inline script in site editor

eri-trabiccolo committed:
start adding templates to the editor

eri-trabiccolo committed:
Improve block templates handling

eri-trabiccolo committed:
fix method name

eri-trabiccolo committed:
Better blocks theme name

eri-trabiccolo committed:
Improve 2022 css

eri-trabiccolo committed:
improve twenty twenty two compatibility

eri-trabiccolo committed:
Allow themes to override llms block templates

eri-trabiccolo committed:
docs

eri-trabiccolo committed:
better block templates titles

eri-trabiccolo committed:
Improve block templates localization

thomasplevy commented on includes/class-llms-block-templates.php:
```diff
@@ -0,0 +1,564 @@
+<?php
+/**
+ * Handles the block templates
```
```suggestion
 * LLMS_Block_Templates class file.
```

Over time I've been switching the way the header matter is defined because the developer docs *do not parse* the file header. Adding a better description to the class header is then output on the parsed site at developer.lifterlms.com

thomasplevy commented on includes/class-llms-block-templates.php:
```diff
@@ -0,0 +1,564 @@
+defined( 'ABSPATH' ) || exit;
+
+/**
+ * LLMS_Block_Templates class.
```
```suggestion
 * Handles the block templates.
```

thomasplevy commented on includes/class-llms-block-templates.php:
```diff
@@ -0,0 +1,564 @@
+	 *
+	 * @var string
+	 */
+	const LLMS_BLOCK_TEMPLATES_NAMESPACE = 'lifterlms/lifterlms';
```
```suggestion
	const LLMS_BLOCK_TEMPLATES_NAMESPACE = 'llms/lifterlms';
```

Add-ons might be `llms/lifterlms-groups` for example?

I think using a shorter version where possible makes sense but maybe this isn't worth the change?

eri-trabiccolo replied:
I used that because (looking at wc) I found that WP automatically turns `lifterlms/lifterlms` (that we use as the `theme` block template property) into LifterLMS (you know, plugin name etc.).
But I think we can, as you propose, go with your suggestion and use this that I called `namespace` to build the id. Then introduce a new constant like:
`const LLMS_BLOCK_TEMPLATES_THEME = 'lifterlms/lifterlms;`
so everybody's happy and makes more sense. Let me check if this doesn't introduce any issue.

eri-trabiccolo replied:
going to revert this because the first part of the id and the "theme name" are somewhat tied, also, what will be saved in the DB (which might occupy more space) is the theme name, and that - for the reason above - I would like to keep it as `lifterlms/lifterlms`.

thomasplevy replied:
Okay! <3

thomasplevy commented on includes/class-llms-block-templates.php:
```diff
@@ -0,0 +1,564 @@
+
+}
+
+$block_templates = new LLMS_Block_Templates();
+$block_templates->init();
+
+return $block_templates;
```
What's the purpose of this new pattern instead of following preexisting conventions?

Most other classes return `new ClassName()` and have their initialization stuff in the `__construct()`, or, in the case of static classes, `return ClassName::init()`

If there's no good reason for this new convention I'd rather change `init()` to `__construct()` and switch the return to follow our existing patterns.

eri-trabiccolo replied:
The purpose of this is that I want to be able to call this object methods from another file without adding the hooks.
See: 
https://github.com/gocodebox/lifterlms/pull/1952/files#diff-7947d18155c2e03348f57a9ef19a46d9e47a3a999f09f521726fb3c2e82a7be7R466

that `init()` should be actually be renamed `hooks()` or something of the sort.
Hooks addition are not `initialization` to me.

If you prefer having them in the constructor I'd say that this must become a singleton, and I add a `block_templates()` method here:
https://github.com/gocodebox/lifterlms/blob/dev/class-lifterlms.php

what do you think?

thomasplevy replied:
Mostly what I don't like here is having functionality / logic buried at the bottom of a class file, especially since we *don't do this anywhere else* it makes it hard to locate in the future since we're not primed to look for any functional code at the bottom of a class file.

I think, based on existing patterns, the singleton makes the most sense then.

eri-trabiccolo replied:
@thomasplevy singleton thing done

thomasplevy commented on includes/class-llms-block-templates.php:
```diff
@@ -0,0 +1,564 @@
+		$block_templates_base_paths = apply_filters(
+			'lifterlms_block_templates_directories',
+			array(
+				LLMS()->plugin_path() . '/templates/' . self::LLMS_BLOCK_TEMPLATES_DIRECTORY_NAME,
```
```suggestion
				llms()->plugin_path() . '/templates/' . self::LLMS_BLOCK_TEMPLATES_DIRECTORY_NAME,
```

thomasplevy commented on includes/class-llms-block-templates.php:
```diff
@@ -0,0 +1,564 @@
+	 * @param string[] $strings Localization admin strings.
+	 * @return string[]
+	 */
+	public function add_block_templates_admin_strings_js( $strings ) {
```
I'd rather code all this stuff into the block directly and rely on `@wordpress/i18n` in favor of using `LLMS.l10n` -- I've been working (silently) to phase out usage of `LLMS.l10n` and I don't see any advantage to doing this this way.

eri-trabiccolo replied:
Well my idea is that this was more easy to extend, in an add-on.
E.g. I'm going to add a block template for the groups plugin, and I wanted to only add the template html + a php file where I add all the filters needed to load the template, and its js title, correctly.
But that's fine I'll add some inline script in groups to do the same (and a filter hook in https://github.com/gocodebox/lifterlms-blocks/pull/147/files#diff-c12382294566f22f66d320592d72c6293bae6b752c5b4eb0296d7990685d81c5R4). As I don't feel to add a js file only for that in groups.

thomasplevy replied:
@eri-trabiccolo I'm find with localizing with PHP instead of JS for lighter-weight extensibility in this case. However, I don't see any reason to use `LLMS.l10n` in order to do so.

Instead, keep this but turn it into a JS localization object and then in JS we can read the map to output the object instead of loading them into `LLMS.l10n` and reading them that way. I think that will work, you?

eri-trabiccolo replied:
so I'm going to localize our blocks script in a more formal way?!

thomasplevy replied:
@eri-trabiccolo I don't think it needs to be anything more formal than outputting an object of translation data, mapping the template ID/name to the translated title, and outputting it as a JS variable using [wp_localize_script()](https://developer.wordpress.org/reference/functions/wp_localize_script/). Add a filter on the object before it's passed into the function. Then add-ons can add their own data to it via PHP.

In the legacy block JS, read the map I hard-coded into the `getLabel()` method from the localized JS object instead of it being hard coded.

eri-trabiccolo replied:
oh, with formal I meant use `wp_localize_script()` yeah :D, rather than build the object as we do in other parts of the codebase. <3

eri-trabiccolo replied:
php-block templates localized

thomasplevy commented on includes/class.llms.template.loader.php:
```diff
@@ -28,9 +28,18 @@ class LLMS_Template_Loader {
 	public function __construct() {
 
+		// Template loading for FSE themes.
+		add_action(
```
I don't want us to start using anonymous functions for hooks.I don't have any really good reason for this but it *feels* like it's a really bad pattern to start implementing, unless there's a really good reason for it. This seems like the reason is "I don't want to add a callback that only adds a filter" and that doesn't feel like a good reason for it in my opinion.

eri-trabiccolo replied:
Well yeah, I thought that isn't a bad pattern and couldn't think why not using it in this specific case. :D 
Why is it a bad pattern? I think it is bad when you might want to unhook a callback, but:
1) in this specific case you wouldn't be able to unhook the `template_redirect` callback from outside this class in any case as the object is instantiated on file loading
2) if it wasn't for 1) you could always unhook the `pre_get_block_templates` callback hooking your "unhooking logic" at `template_redirect|11`.
That was my reasoning: no reason to not do that, and gold.

That said, I'm going to follow your feeling <3

thomasplevy commented on includes/class.llms.template.loader.php:
```diff
@@ -462,52 +516,62 @@ public function template_loader( $template ) {
+		}
 
-				return $template;
+		$forced_template = apply_filters( 'llms_force_llms_template_loading', true ) ? $this->get_maybe_forced_llms_template() : false;
```
```suggestion
		$forced_template = apply_filters( 'llms_force_template_loading', true ) ? $this->get_maybe_forced_llms_template() : false;
```

Also needs docblock

thomasplevy commented on includes/class.llms.template.loader.php:
```diff
@@ -462,52 +516,62 @@ public function template_loader( $template ) {
+	 *
+	 * @return null|string
+	 */
+	private function get_maybe_forced_llms_template() {
```
```suggestion
	private function get_maybe_forced_template() {
```

Is `llms` necessary in our own internal private function?

eri-trabiccolo replied:
wasn't a prefixing choice, it was more a naming choice, a name that tells we're talking about llms specific templates.

thomasplevy commented on includes/functions/llms.functions.template.php:
```diff
@@ -254,3 +255,22 @@ function llms_get_template_override_directories() {
+
+	// We have reason to use a LifterLMS template, check if there's an override we should use from a theme / etc...
+	$override      = llms_get_template_override( $template, $extension );
+	$template_path = $override ? $override : LLMS()->plugin_path() . "/{$template_directory}/";
```
```suggestion
	$template_path = $override ? $override : llms()->plugin_path() . "/{$template_directory}/";
```

thomasplevy commented on includes/functions/llms.functions.template.php:
```diff
@@ -193,23 +193,24 @@ function llms_locate_template( $template_name, $template_path = '', $default_pat
  *
  * @since Unknown
  * @since 4.8.0 Move template override directories logic into llms_get_template_override_directories.
+ * @since [version] Added `$extension`, and `$is_block_template` parameter.
```
```suggestion
 * @since [version] Added `$extension` parameter.
```

thomasplevy commented on includes/functions/llms.functions.template.php:
```diff
@@ -193,23 +193,24 @@ function llms_locate_template( $template_name, $template_path = '', $default_pat
  * @return mixed Template file or false if none exists.
  */
-function llms_get_template_override( $template = '' ) {
+function llms_get_template_override( $template = '', $extension = '' ) {
```
This means that block templates and php templates can both be stored in the same override directory, right? 

A new block template (in LifterLMS) is stored in `templates/block-templates/archive-course.html` and a php template could be at `templates/archive-course.php`

So if a theme wanted to override both of these, the directory would have them at `lifterlms/archive-course.html` and `lifterlms/archive-course.php` (both in the same folder)

Right?

I think this is fine but I just want to make sure I'm reading this right.

eri-trabiccolo replied:
Nope, themes have to override the block templates in
`lifterlms/block-templates/`

The reason for `$extension` is that... I think we just don't need it anymore. I added it when I was doing something else, but it's not needed anymore (gonna check). I'll remove it. :D

thomasplevy commented on templates/block-templates/single-certificate.html:
```diff
@@ -0,0 +1,2 @@
+
```
Is the leading space needed for some reason?

eri-trabiccolo replied:
no, typo, thanks.

thomasplevy commented on templates/block-templates/single-no-access.html:
```diff
@@ -0,0 +1,5 @@
+<!-- wp:template-part {"slug":"header"} /-->
+<!-- wp:group {"layout":{"inherit":true}} -->
+<div class="wp-block-group"><!-- wp:llms/php-template {"template":"single-no-access", "title":"LifterLMS Single Template requiring Membership"} /--></div>
```
```suggestion
<div class="wp-block-group"><!-- wp:llms/php-template {"template":"single-no-access", "title":"LifterLMS Template Access Restricted"} /--></div>
```

This is not often displayed but I don't think it's displayed exclusively when membership is required so I think the title should reflect that to prevent confusion or bug reports or whatever.

eri-trabiccolo replied:
mmm I think you're right :D

thomasplevy commented:
@eri-trabiccolo overall I think this is really solid. I've tested it on 5.9-rc4 and haven't spotted any show-stopping issues

However, when loading the membership catalog and single certificate templates (in the site editor) the php-template block renders no title:

![Screenshot_2022-01-24_16-27-07](https://user-images.githubusercontent.com/1290739/150882377-0cb90d00-81be-400c-9efe-78b5ad11259c.png)
![Screenshot_2022-01-24_16-27-54](https://user-images.githubusercontent.com/1290739/150882379-ad8ad688-243b-4128-974d-b89c15a61ad3.png)

eri-trabiccolo committed:
Apply suggestions from code review

Co-authored-by: Thomas Patrick Levy <thomas@gocodebox.com>

eri-trabiccolo commented:
> @eri-trabiccolo overall I think this is really solid. I've tested it on 5.9-rc4 and haven't spotted any show-stopping issues
> 
> However, when loading the membership catalog and single certificate templates (in the site editor) the php-template block renders no title:
> 
> ![Screenshot_2022-01-24_16-27-07](https://user-images.githubusercontent.com/1290739/150882377-0cb90d00-81be-400c-9efe-78b5ad11259c.png) ![Screenshot_2022-01-24_16-27-54](https://user-images.githubusercontent.com/1290739/150882379-ad8ad688-243b-4128-974d-b89c15a61ad3.png)

Oh thanks for reporting! It's because of the double closing brace on this line:
https://github.com/gocodebox/lifterlms/pull/1952/files#diff-8fc257ecd0c4769c75c10a3a7fe9b40a8b519caa0f9ab5b69a5d182055dfe6caR3

:D

eri-trabiccolo committed:
Fix typos in block templates and update single no access title

eri-trabiccolo committed:
revert LLMS_BLOCK_TEMPLATE_NAMESPACE as lifterlms/lifterlms

eri-trabiccolo committed:
Remove file extension parameter for some template functions - not needed

eri-trabiccolo committed:
fix cs

eri-trabiccolo committed:
Remove php block templates title attributes, they will be handled via js params localization

eri-trabiccolo committed:
LLMS_Block_Templates is now a singleton
and is able to handle block templates from add-ons when oppropriately configured
php-template block titles in the block edior are now handled via script localization

eri-trabiccolo committed:
fix cs

eri-trabiccolo commented:
@thomasplevy wanna have another look at this?

thomasplevy commented on includes/class-llms-block-templates.php:
```diff
@@ -0,0 +1,663 @@
+	public function localize_blocks() {
+		wp_localize_script(
+			'llms-blocks-editor',
+			'LLMS_Block_Templates_l10n',
```
I know that in a lot of the codebase we use phpcs standards for JS and this kind of stuff but as since this will ultimately be a JS variable we should make it a JS-style variable.

```suggestion
			'llmsBlockTemplatesL10n',
```

Commit:
Merge branch 'dev' into issue/1824

Commit:
Update includes/class-llms-block-templates.php
